src/lxml/serializer.pxi

```diff
@@ -1035,3 +1044,22 @@ cdef class _FileWriterElement:
     def __exit__(self, exc_type, exc_val, exc_tb):
         self._writer._write_end_element(self._element)
         self._writer._method = self._old_method
+
+@cython.final
+@cython.internal
+@cython.freelist(8)
+cdef class _MethodChanger:
+    cdef int _new_method
+    cdef int _old_method
+    cdef _IncrementalFileWriter _writer
+
+    def __cinit__(self, _IncrementalFileWriter writer not None, int method):
+        self._writer = writer
+        self._new_method = method
+        self._old_method = writer._method
```

Comment:
I actually don't think that we need a stack here. It should be enough to check that the output method didn't change between the end of `__enter()__` and the start of `__exit()__`, i.e. to save the old method on entry and check that the current method is still `new_method` on exit. If it's not, it's a bug in the calling code and the user should know about it, i.e. get an exception. Oh, and maybe prevent re-entry before exiting by checking that `old_method` is set to an invalid value on entry. That would detect programming problems while not restricting users too much.